### PR TITLE
Only visit reachable blocks in ConstProp lint.

### LINKED ISSUE #78803: Overzealous unconditional_panic (denied) warning
To try and workaround the lack of `const` `Result::unwrap` on 1.47 I implemented a basic variant of it manually, but this fails to compile without suppressing the `unconditional_panic` warning:

```rust
#[derive(Copy, Clone)]
pub struct Foo;

impl Foo {
    pub const fn new_unwrap(success: bool) -> Self {
        let result = if success { Ok(Foo) } else { Err(()) };
        [][match result {
            Ok(foo) => return foo,
            Err(_) => 0,
        }]
    }
}

fn main() {}
```

```
error: this operation will panic at runtime
  --> uwu.rs:7:9
   |
7  | /         [][match result {
8  | |             Ok(foo) => return foo,
9  | |             Err(_) => 0,
10 | |         }]
   | |__________^ index out of bounds: the length is 0 but the index is 0
   |
   = note: `#[deny(unconditional_panic)]` on by default

error: aborting due to previous error
```

This code panicking _is_ conditional on the input `success` value.

@rustbot modify labels: +A-const-fn, +A-lint

### LINKED ISSUE #109731: const-propagated `arithmetic_overflow` in unreachable code
### Code

I tried this code: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f6804071f4851d138db85474dbd096c8)

```rust
fn main() {
    let x = 2u32;
    let y = 3u32;
    // ...
    if y <= x {
        dbg!(x - y);
    }
}
```

I expected to see this happen: compiles, nothing printed

Instead, this happened:

```
error: this arithmetic operation will overflow
 --> src/main.rs:6:14
  |
6 |         dbg!(x - y);
  |              ^^^^^ attempt to compute `2_u32 - 3_u32`, which would overflow
  |
  = note: `#[deny(arithmetic_overflow)]` on by default
```

The original code comes from macro-generated test cases in `num-integer`, which failed here:
https://github.com/rust-num/num-integer/pull/54#issuecomment-1488962831

### Version it worked on

It most recently worked on: 1.69.0-beta.3

### Version with regression

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (478cbb42b 2023-03-28)
binary: rustc
commit-hash: 478cbb42b730ba4739351b72ce2aa928e78e2f81
commit-date: 2023-03-28
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.0
```

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/78803
Fixes https://github.com/rust-lang/rust/issues/109731

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Does this also fix #78803?

**Comment by cjgillot** [comment]
Yes, added a test.

File: compiler/rustc_mir_transform/src/const_prop_lint.rs [hunk 1]
```diff
@@ -649,6 +661,17 @@ impl<'tcx> Visitor<'tcx> for ConstPropagator<'_, 'tcx> {
...
+            TerminatorKind::SwitchInt { ref discr, ref targets } => {
+                if let Some(ref value) = self.eval_operand(&discr, location)
+                  && let Some(value_const) = self.use_ecx(location, |this| this.ecx.read_scalar(&value))
+                  && let Ok(constant) = value_const.try_to_int()
+                  && let Ok(constant) = constant.to_bits(constant.size())
+                {
+                    let target = targets.target_for_value(constant);
+                    self.worklist.push(target);
+                    return;
+                }
```
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/const_prop_lint.rs]: leave a comment here stating that if we don't know the discriminant, we assume all targets are reachable below.

**Comment by cjgillot** [comment]
Just in case:
@bors try @rust-timer queue

**Comment by oli-obk** [comment]
r=me after a rebase

**Comment by apiraino** [comment]
Going to nominate for a backport because one issue fixed is a `P-high`

@rustbot label +beta-nominated

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-04-06/near/347432330)

@rustbot label +beta-accepted

**Comment by apiraino** [comment]
this was not fixing a beta regression (my bad for misunderstanding the context)

@rustbot label -beta-accepted -beta-nominated

**Pull request merged**